Draft implementation of diagnose event names

hazzadous opened the pull request:
## Changes

Add endpoint for getting correlation values between events and funnel success.

## How did you test this code?

*Please describe.*

Commit:
chore(diagnose): add a stub endpoint for event diagnosis

This adds an insight endpoint that takes a `target_entity` and returns
a list of events ordered by significance of a person reaching
`target_entity`

Followup PRs will add actual calculations but this should act as a
conversation piece around the structure of request and response, as well
as a stub from which UI development can start.

Commit:
play around to get mypy typings right

Commit:
feat(diagnose): add diagnose stubs for implementation

Commit:
Sort out test reponse structurea

neilkakkar commented on ee/clickhouse/queries/diagnose/diagnose.py:
```diff
@@ -11,4 +19,76 @@ def __init__(self, filter: DiagnoseFilter, team: Team) -> None:
+           100 different names for events. We should avoid n+1 queries for the
+           event names dimension
+
+        Contincency tables are something we can pull out of the db, so we can
```
This is AMAZING. Thanks for writing this up! :D

And yes, this approach makes sense, with a caveat:

It's not just cohort criteria, but a funnel step, or a Paths step that can determine these people. Right now, we're only dealing with funnels, so makes sense to not-generalise, since we're not yet sure how things will look for the rest.

(Cohort and Funnel step have different semantics: A cohort is just a list of people, while a funnel step is a list of people + specific timestamps in their journey through the funnel - so you literally need the funnel query to get the people & their timestamps)


Also note that, we're not just looking for success, but failure odds too! (It can make sense to separate these out when dealing with too many event names - so you get the top of the list for success counts, and bottom of the list for failure counts, instead of getting the entire thing and trying to sort in Python). Worth adding to requirements.


```suggestion
        Contingency tables are something we can pull out of the db, so we can
```

neilkakkar commented on ee/clickhouse/queries/diagnose/diagnose.py:
```diff
@@ -3,6 +3,14 @@
+    return 1
+
+
 class ClickhouseDiagnose:
```
Diagnose is actually the overall theme. Diagnosing causes includes: the funnel to Paths setup, improving Paths, Quant analysis, and correlation analysis!

I'd propose: (1) Calling the folder quant_analysis & (2) Calling this specific class: `ClickhouseCorrelationAnalysis`

neilkakkar commented on ee/clickhouse/queries/diagnose/diagnose.py:
```diff
@@ -3,6 +3,14 @@
+
+
 class ClickhouseDiagnose:
     _filter: DiagnoseFilter
```
We're trying to move towards unifying filters. Specially, since this filter (limited to funnels) uses the same filter as funnels, makes sense to use `Filter` here as well? We'll add the 2 new properties to `Filter` itself

Commit:
refactor: address CR comments

Commit:
feat(funnel): add event correlation calculation implementation

This adds to the `/api/projects/<team_id>/funnel/correlation` endpoint
an implementation that calculates the odds_ratios for each event that a
user that has been part of a funnel, successful or otherwise.

neilkakkar commented on ee/api/test/test_funnel_correlation.py:
```diff
@@ -1,9 +1,14 @@
-from typing import Any, Dict
+import dataclasses
```
A file here instead: https://github.com/PostHog/posthog/blob/master/ee/clickhouse/views/test/test_clickhouse_insights.py#L59 seems like a good place for these tests

Commit:
chore(correlation): get query working

Commit:
refactor(correlation): move functions to Query methods

neilkakkar commented on ee/clickhouse/views/insights.py:
```diff
@@ -110,6 +111,29 @@ def calculate_funnel(self, request: Request) -> Dict[str, Any]:
             return {"result": funnel_order_class(team=team, filter=filter).run()}
 
+    # ******************************************
+    # /insight/funnel_correlation
```
```suggestion
    # /insight/funnel/correlation
```

neilkakkar commented on ee/clickhouse/views/insights.py:
```diff
@@ -110,6 +111,29 @@ def calculate_funnel(self, request: Request) -> Dict[str, Any]:
+    # Returns significant events, i.e. those that are correlated with a person
+    # making it through a funnel
+    # ******************************************
+    @action(methods=["GET"], url_path="funnel/correlation", detail=False)
```
```suggestion
    @action(methods=["POST"], url_path="funnel/correlation", detail=False)
```

neilkakkar replied:
For follow up PR, not here!
